lib/network/rest/server.js

```diff
@@ -0,0 +1,332 @@
+            level: config.logLevel,
+            dump: config.dumpLevel,
+        };
+    }
+    return new Logger('RESTServer', options);
+}
+
+/**
+ * @class
+ * @classdesc REST Server interface
+ *
+ * You have to call setup() to initialize the storage backend, then
+ * start() to start listening to the configured port.
+ */
+class RESTServer extends httpServer {
+
+    /**
+     * @constructor
+     * @param {Object} params - constructor params
+     * @param {Number} params.port - TCP port where the server listens to
+     * @param {Object} [params.log] - logger configuration
+     */
+    constructor(params) {
+        assert(params.port);
+
+        const logging = createLoggingObject(params.log);
+        super(params.port, logging);
+        this.logging = logging;
+        this.dataStore = params.dataStore;
+
+        this.onRequest(this._onRequest);
+    }
+
+    /**
+     * Setup the storage backend
+     *
+     * @param {function} callback - called when finished
+     * @return {undefined}
+     */
+    setup(callback) {
+        this.dataStore.setup(callback);
+    }
+
+    /**
+     * Create a new request logger object
+     *
+     * @param {String} reqUids - serialized request UIDs (as received in
+     * the X-Scal-Request-Uids header)
+     * @return {Logger} new request logger
+     */
+    createLogger(reqUids) {
+        return reqUids ?
+            this.logging.newRequestLoggerFromSerializedUids(reqUids) :
+            this.logging.newRequestLogger();
+    }
+
+    /**
+     * Main incoming request handler, dispatches to method-specific
+     * handlers
+     *
+     * @param {http.IncomingMessage} req - HTTP request object
+     * @param {http.ServerResponse} res - HTTP response object
+     * @return {undefined}
+     */
+    _onRequest(req, res) {
+        const reqUids = req.headers['x-scal-request-uids'];
+        const log = this.createLogger(reqUids);
+        log.debug('request received', { method: req.method,
+                                        url: req.url });
+        const reqMethods = {
+            PUT: this._onPut.bind(this),
+            GET: this._onGet.bind(this),
+            DELETE: this._onDelete.bind(this),
+        };
+        if (req.method in reqMethods) {
+            reqMethods[req.method](req, res, log);
+        } else {
+            // Method Not Allowed
+            sendError(res, 405);
+        }
+    }
+
+    /**
+     * Handler for PUT requests
+     *
+     * @param {http.IncomingMessage} req - HTTP request object
+     * @param {http.ServerResponse} res - HTTP response object
+     * @param {Logger} log - logger object
+     * @return {undefined}
+     */
+    _onPut(req, res, log) {
+        const urlObj = url.parse(req.url);
+        const pathInfo = utils.explodePath(urlObj.path);
+        if (!pathInfo) {
+            return sendError(res, 403, 'invalid URL');
+        }
+        if (pathInfo.service !== constants.dataFileURL) {
+            return sendError(res, 403,
```

Comment:
`sendError` is actually forging the HTTP response, nothing more, it does not actually create an Error object.

I see a value of using arsenal errors when they match a specific error type. In this case, as an example, no arsenal error tells that the user has entered an invalid URI, so it turns that either I have to create a new arsenal error containing this message and a 403 error code (there are already a dozen, most S3-oriented), or I have to return an "AccessDenied" message of some sort coming from an existing error, that does not contain the reason, which can make it harder to debug invalid use. Though maybe it can be used to reconstruct the matching arsenal error on client side, but only if we parse the description, not sure it's practical to do.